## Fibonacci 

In [73]:
import sympy as sym
import math
import timeit

In [74]:
l=sym.Symbol('l')


In [75]:
n_list=[0,1,2,3,4,5,6,7,8,9,10]
f_n_list=[1,1,2,3,5,8,13,21,34,55,89]

In [76]:
def calculate_f_n(b,a,epsilon):
    f_n=(b-a)/epsilon
    return f_n

def calculate_n(length_fibonacci_list,f_n):
    for i in range(length_fibonacci_list):
        if(f_n <=  f_n_list[i] ):
                f_n=f_n_list[i]
                n=n_list[i]
                break
    return f_n,n

def calculate_r_k(n):
    r_k=[]
    for k in range (n):
        r_k.append(f_n_list[n-k-1]/f_n_list[n-k])
    return r_k  

def calculate_lambda(b,a,mini,n,r_k):        # In case of minimization set mini flag to 1 , maximization set mini flag to 0
    org_a=a
    org_b=b
    for k  in range (n-1):

        d=(b-a)*r_k[k]

        x_1=b-d

        f_x_1=f_l.evalf(subs={l:x_1})

        x_2=a+d  

        f_x_2=f_l.evalf(subs={l:x_2})  

        check=f_x_2 >= f_x_1

        case=check^mini
        if (case==False):
            a=a 
            b=x_2

        elif (case==True):
            print(str(k)+"I am here ")
            a=x_1
            b=b
        else:

            print("unhandled case")
            
    estimated_lamda=round((b+a)/2,5)
    return estimated_lamda,a,b

def calculate_percentage_reduction_ratio(org_b,org_a,a,b):
    reduction_ratio=round(((b-a)/(org_b-org_a))*100,2)
    return reduction_ratio

In [77]:
def fiboancci_main(org_b,org_a,epsilon,length_fibonacci_list,mini):
    
    f_n=calculate_f_n(org_b,org_a,epsilon)
    
    f_n,n=calculate_n(length_fibonacci_list,f_n)
    
    r_k=calculate_r_k(n)
    
    estimated_lamda,a,b=calculate_lambda(org_b,org_a,mini,n,r_k)
    
    reduction_ratio=calculate_percentage_reduction_ratio(org_b,org_a,a,b)
    return estimated_lamda,reduction_ratio,n

### Example to be solved using code 

<img src="images/fibonacci.png">

<img src="images/fibonacci2.png">

In [78]:
n_list=[0,1,2,3,4,5,6,7,8,9,10]
f_n_list=[1,1,2,3,5,8,13,21,34,55,89]
length_fibonacci_list=len(f_n_list)



## How to test Fibonacci code ?

In [79]:
sfTime=timeit.default_timer()

f_l=25600*l**4-25600*l**3+6416*l**2-16*l+4 # f(lambda)

org_b=0.1                                  # end of region b , org stands for original i.e. b value at the end of region
org_a=0                                    # start of region a 
epsilon=0.01                            
       
mini=1                                    # mini stands for minimization if problem of type minimization it will equal 1 
                                          # if not it will equal to zero 
estimated_lamda,reduction_ratio,n=fiboancci_main(org_b,org_a,epsilon,length_fibonacci_list,mini)

fTime=timeit.default_timer() - sfTime

### Analysis 

In [80]:
estimated_lamda

0.00385

In [81]:
reduction_ratio

7.69

In [82]:
optimal_value=f_l.evalf(subs={l:estimated_lamda})


In [83]:
print ("Fibonacci : Number of iterations",n)
print ("Fibonacci : Optimal solution /Lambda star",estimated_lamda)
print("Fibonacci :  Optimal value",optimal_value)
print ("Fibonacci : CPU Time",fTime)

Fibonacci : Number of iterations 6
Fibonacci : Optimal solution /Lambda star 0.00385
Fibonacci :  Optimal value 4.03204587888656
Fibonacci : CPU Time 0.02207130199894891


## Golden section 

In [84]:
def golden_calculate_n(b,a,epi):
    e=math.log(epi/(b-a),10)
    f=math.log(0.618,10)
    n=math.ceil((e/f)+1 )
    return n 

In [85]:

def golden_calculate_lambda(b,a,mini,n,r_k=0.618):        # In case of minimization set mini flag to 1 , maximization set mini flag to 0
    org_a=a
    org_b=b
    for k  in range (n-1):

        d=(b-a)*r_k

        x_1=b-d

        f_x_1=f_l.evalf(subs={l:x_1})

        x_2=a+d  

        f_x_2=f_l.evalf(subs={l:x_2})  

        check=f_x_2 >= f_x_1

        case=check^mini
        if (case==False):
            a=a 
            b=x_2

        elif (case==True):
            print(str(k)+"I am here ")
            a=x_1
            b=b
        else:

            print("unhandled case")
            
    estimated_lamda=round((b+a)/2,5)
    return estimated_lamda,a,b


In [86]:
def golden_main(org_b,org_a,epi,mini):
    
    n=golden_calculate_n(org_b,org_a,epi)
    
    estimated_lamda,a,b=golden_calculate_lambda(org_b,org_a,mini,n,r_k=0.618)
    reduction_ratio=calculate_percentage_reduction_ratio(org_b,org_a,a,b)
    return estimated_lamda,reduction_ratio,n

In [87]:
golden_startTime=timeit.default_timer() 

## How to test Golden section code ?

In [88]:

f_l=25600*l**4-25600*l**3+6416*l**2-16*l+4
org_b=0.1
org_a=0
epsilon=0.01
mini=1
estimated_lamda,reduction_ratio,n=golden_main(org_b,org_a,epsilon,mini)
print("estimated_lamda ",estimated_lamda)
print("reduction_ratio  ",reduction_ratio)

estimated_lamda  0.00451
reduction_ratio   9.01


In [89]:
golden_Time=timeit.default_timer() -golden_startTime

### Calculations 


In [90]:
optimal_value=f_l.evalf(subs={l:estimated_lamda})

In [91]:
print ("Golden Number of iterations",n)
print ("Golden Optimal solution /Lambda star",estimated_lamda)
print("Golden Optimal value",optimal_value)
print ("Golden CPU Time",golden_Time)

Golden Number of iterations 6
Golden Optimal solution /Lambda star 0.00451
Golden Optimal value 4.05600428623790
Golden CPU Time 0.2974034020007821


## Quadratic interpolation 

In [92]:
def calculate_lambda_star_first_refit(fA,fB,fC,t_o):
    s_lambdaa=((4*fB-3*fA-fC)/(4*fB-2*fC-2*fA))*t_o
    return s_lambdaa 
def calculate_a_b_c_first_refit(fA,fB,fC,t_o):
    a = fA
    b = (4*fB - 3*fA - fC)/(2*t_o)
    c = (fC + fA - 2*fB)/(2*(t_o)**2)
    return a, b,c 

In [93]:
def first_refit(f1,fA,t_o):
    if (f1>fA):

        fC=f1

        fB=f_l.evalf(subs={l:t_o/2})

        t_o=t_o/2

        s_lambdaa=calculate_lambda_star_first_refit(fA,fB,fC,t_o)

        a,b,c=calculate_a_b_c_first_refit(fA,fB,fC,t_o)
        A=0
        B=t_o
        C=2*t_o
    if (f1<=fA) :

        fB=f1

        f2=f_l.evalf(subs={l:t_o*2})
        if (f2>f1):
            fC=f2
            s_lambdaa=calculate_lambda_star_first_refit(fA,fB,fC,t_o)

            a,b,c=calculate_a_b_c_first_refit(fA,fB,fC,t_o)
            A=0
            B=t_o
            C=2*t_o

        else :
            while (f2<f1):
                f1=f2
                t_o=2*t_o

                if (f1>fA):

                    fC=f1

                    fB=f_l.evalf(subs={l:t_o/2})

                    t_o=t_o/2

                    s_lambdaa=calculate_lambda_star_first_refit(fA,fB,fC,t_o)

                    a,b,c=calculate_a_b_c_first_refit(fA,fB,fC,t_o)
                    break
                if (f1<=fA) :

                    fB=f1

                    f2=f_l.evalf(subs={l:t_o*2})
                    if (f2>f1):
                        fC=f2
                        s_lambdaa=calculate_lambda_star_first_refit(fA,fB,fC,t_o)

                        a,b,c=calculate_a_b_c_first_refit(fA,fB,fC,t_o)
                        break 

        A=0
        B=t_o
        C=2*t_o
        
        fs_lambdaa=f_l.evalf(subs={l:s_lambdaa}) 
        
        return A,fA,B,fB,C,fC,s_lambdaa,fs_lambdaa,a,b,c  
        
            
                

In [94]:
def test_convergence(epi_one,s_lambdaa,a,b,c):
    
    s_h=a+b*s_lambdaa+c*s_lambdaa**2 # h at lambda star 
    
    s_f=f_l.evalf(subs={l:s_lambdaa})
    
    ratio =abs ((s_h-s_f)/s_f)
    if (ratio <=epi_one ):
        acceptable=1
    else:
        acceptable=0
    return acceptable , ratio

In [95]:
def calculate_lambda_a_b_c(A,fA,B,fB,C,fC) :
    lambdaa = (fA*(B**2-C**2)+fB*(C**2-A**2)+fC*(A**2-B**2))/(2*(fA*(B-C)+fB*(C-A)+fC*(A-B)))
    a = (fA*B*C*(C-B)+fB*C*A*(A-C)+fC*A*B*(B-A))/((A-B)*(B-C)*(C-A))
    b = (fA*(B**2-C**2)+fB*(C**2-A**2)+fC*(A**2-B**2))/((A-B)*(B-C)*(C-A))
    c = -1*(fA*(B-C)+fB*(C-A)+fC*(A-B))/((A-B)*(B-C)*(C-A))
    return lambdaa, a, b ,c 
 

In [96]:
# lambda here is lambda star in previous refits 
# f is f_star 
def additional_refits(A,fA,B,fB,C,fC,lambdaa,f):

# Table 5.5 refining scheme from rao 
        if (lambdaa > B) and (f < fB):
            A = B
            B = lambdaa
            fA = fB
            fB = f
        if (lambdaa > B) and (f > fB):
            C = lambdaa
            fC = f
        if (lambdaa < B) and (f < fB):
            C = B
            B = lambdaa
            fB = fC
            fB = f
        if (lambdaa < B) and (f > fB):
            A = lambdaa
            fA = f
        s_lambdaa,a,b,c=calculate_lambda_a_b_c(A,fA,B,fB,C,fC)
        
        fs_lambdaa=f_l.evalf(subs={l:s_lambdaa})   
        
        return  A,fA,B,fB,C,fC,s_lambdaa,fs_lambdaa,a,b,c  

In [97]:
def quadratic_interpolation_main(f_l,t_o,epi_one):  
    
    fA=f_l.evalf(subs={l:0})
    
    f1=f_l.evalf(subs={l:t_o})
    
    A,fA,B,fB,C,fC,s_lambdaa,fs_lambdaa,a,b,c =first_refit(f1,fA,t_o)
    
    acceptable , ratio=test_convergence(epi_one,s_lambdaa,a,b,c)
    counter=1
    if (acceptable ==1):      
                      
        return s_lambdaa,fs_lambdaa,counter
    else :
        while (acceptable==0):
                  
            A,fA,B,fB,C,fC,s_lambdaa,fs_lambdaa,a,b,c  =additional_refits(A,fA,B,fB,C,fC,s_lambdaa,fs_lambdaa)
            
            acceptable=test_convergence(epi_one,s_lambdaa,a,b,c)
            counter=counter+1
        return s_lambdaa,fs_lambdaa,counter
    

## How to test quadratic interpolation code ?


<img src="images/quadratic_interpolation.png">

In [101]:
l=sym.Symbol('l')
f_l=l**5-5*l**3-20*l+5
t_o=0.5
quadratic_interpolation_main(f_l,t_o,1)

(1.65845810225057, -38.4304704180808, 2)

#### This means lambda star equals 1.658 , f_val=-38.43, number of iterations =2

### How to test quadratic interpolation 

In [102]:
q_I_startTime=timeit.default_timer() 

f_l=25600*l**4-25600*l**3+6416*l**2-16*l+4
t_o = 0.001
s_lambdaa,fs_lambdaa,counter=quadratic_interpolation_main(f_l,t_o,1)

    
q_I_time =timeit.default_timer()- q_I_startTime

### This means lambda star equals 0.001258 , f_val=3.98997, number of iterations =1


In [103]:
print ("Quadratic interpolation : Number of iterations",counter)
print ("Quadratic interpolation : Optimal solution /Lambda star",s_lambdaa)
print("Quadratic interpolation :  Optimal value",fs_lambdaa)
print ("Quadratic interpolation : CPU Time",q_I_time)

Quadratic interpolation : Number of iterations 1
Quadratic interpolation : Optimal solution /Lambda star 0.00125792708535247
Quadratic interpolation :  Optimal value 3.98997482706145
Quadratic interpolation : CPU Time 0.0046636689985462
